builtin/providers/vsphere/resource_vsphere_virtual_machine.go

```diff
@@ -0,0 +1,1062 @@
+package vsphere
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"golang.org/x/net/context"
+)
+
+var DefaultDNSSuffixes = []string{
+	"vsphere.local",
+}
+
+var DefaultDNSServers = []string{
+	"8.8.8.8",
+	"8.8.4.4",
+}
+
+type networkInterface struct {
+	deviceName  string
+	label       string
+	ipAddress   string
+	subnetMask  string
+	adapterType string // TODO: Make "adapter_type" argument
+}
+
+type hardDisk struct {
+	size int64
+	iops int64
+}
+
+type virtualMachine struct {
+	name              string
+	datacenter        string
+	cluster           string
+	resourcePool      string
+	datastore         string
+	vcpu              int
+	memoryMb          int64
+	template          string
+	networkInterfaces []networkInterface
+	hardDisks         []hardDisk
+	gateway           string
+	domain            string
+	timeZone          string
+	dnsSuffixes       []string
+	dnsServers        []string
+}
+
+func resourceVSphereVirtualMachine() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceVSphereVirtualMachineCreate,
+		Read:   resourceVSphereVirtualMachineRead,
+		Delete: resourceVSphereVirtualMachineDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"vcpu": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"memory": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"datacenter": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"cluster": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"resource_pool": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"gateway": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"domain": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "vsphere.local",
+			},
+
+			"time_zone": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "Etc/UTC",
+			},
+
+			"dns_suffix": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				ForceNew: true,
+			},
```

Comment:
`TypeList` is for attributes where the order of the values is significant. If order does not matter, `TypeSet` is what you want. I'm not familiar with the upstream APIs so this may be correct, but I wanted to let you know so you can double check this for each of your `TypeList` attributes.